File: tools/discover.js
'use strict';

// Finding the set, so nobody has to go looking in their router: every Samsung television answers
// http://<ip>:8001/api/v2/ with its model, and the local /24 is one sweep.

const net = require('net');
const { networkInterfaces } = require('os');
const { createInterface } = require('readline');

const ui = require('./ui.js');
const { describe, DEVICE_API_PORT } = require('./tv.js');

// A /24's worth in flight at once, so each subnet costs about one timeout rather than several.
const PROBE_TIMEOUT = 1500;
const AT_ONCE = 254;

const IPV4 = /^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/;

const isAddress = (text) => {
    const match = IPV4.exec(text);
    return Boolean(match) && match.slice(1).every((part) => Number(part) < 256);
};

const asNumber = (ip) => ip.split('.').reduce((total, part) => (total * 256) + Number(part), 0);

// Only the /24 each interface sits in: a /16 is 65k probes, and a set on a different one is
// something the operator can type.
const subnets = () => {
    const interfaces = networkInterfaces();
    const prefixes = [];
    const mine = [];

    for (const name in interfaces) {
        for (const entry of interfaces[name] || []) {
            if (entry.family !== 'IPv4' || entry.internal) continue;

            mine.push(entry.address);

            const prefix = entry.address.split('.').slice(0, 3).join('.');
            if (prefixes.indexOf(prefix) === -1) prefixes.push(prefix);
        }
    }

    return { prefixes, mine };
};

const pool = async (items, worker) => {
    const found = [];
    let next = 0;

    const run = async () => {
        while (next < items.length) {
            const at = next;
            next += 1;

            const result = await worker(items[at]);
            if (result) found.push(result);
        }
    };

    const runners = [];
    for (let n = 0; n < Math.min(AT_ONCE, items.length); n += 1) runners.push(run());

    await Promise.all(runners);

    return found;
};

// fetch costs far more than the network does: a /24 asked with it takes ten seconds where the same
// range knocked on by hand takes one and a half. So the sweep opens a socket at every address and
// only asks the few that answered what they are.
const knock = (ip) => new Promise((answered) => {
    const socket = net.connect({ host: ip, port: DEVICE_API_PORT });

    const settle = (open) => {
        socket.destroy();
        answered(open ? ip : null);
    };

    socket.setTimeout(PROBE_TIMEOUT, () => settle(false));
    socket.on('connect', () => settle(true));
    socket.on('error', () => settle(false));
});

const sweep = async () => {
    const { prefixes, mine } = subnets();

    const addresses = prefixes.reduce((all, prefix) => {
        for (let host = 1; host < 255; host += 1) {
            const ip = `${prefix}.${host}`;
            if (mine.indexOf(ip) === -1) all.push(ip);
        }
        return all;
    }, []);

    // Something else may sit on 8001, so the answer still has to look like a television.
    const listening = await pool(addresses, knock);

    const televisions = (await Promise.all(listening.map(async (ip) => {
        const device = await describe(ip);
        return device ? { ip, device } : null;
    }))).filter(Boolean);

    return {
        prefixes,
        televisions: televisions.sort((a, b) => asNumber(a.ip) - asNumber(b.ip))
    };
};

const ENTITIES = { quot: '"', apos: '\'', '#39': '\'', amp: '&', lt: '<', gt: '>' };

// A set named `65" OLED` reports it as `65&quot; OLED`.
const plain = (text) => String(text).replace(/&(#39|quot|apos|amp|lt|gt);/g, (whole, name) => ENTITIES[name] || whole);

const nameOf = (device) => plain(device.name || device.modelName || device.model || 'Samsung TV');

// Developer Mode is the next thing that would fail, so it is worth saying before anything is chosen.
const modeOf = (device) => (device.developerMode === '1' ? 'developer mode on' : 'developer mode OFF');

// The model tells two sets with the same name apart, and it is what a forum post would ask for.
const detailOf = (device) => [device.modelName, modeOf(device)].filter(Boolean).join(' \u00b7 ');

const askOnce = (rl, text) => new Promise((resolve) => {
    const closed = () => resolve(null);

    rl.once('close', closed);
    rl.question(text, (answer) => {
        rl.removeListener('close', closed);
        resolve(answer);
    });
});

const listed = (televisions) => {
    televisions.forEach((found, at) => {
        const number = ui.style.bold(String(at + 1).padStart(3));
        ui.note(`${number}  ${nameOf(found.device).padEnd(24)} ${found.ip.padEnd(16)} ${ui.style.dim(detailOf(found.device))}`);
    });
};

const ask = async (televisions) => {
    const rl = createInterface({ input: process.stdin, output: process.stdout });

    const prompt = televisions.length > 1
        ? `Which one? 1-${televisions.length}, or an address: `
        : (televisions.length === 1 ? 'Enter to use it, or another address: ' : 'Address of the TV: ');

    try {
        for (;;) {
            const answer = await askOnce(rl, `\n${prompt}`);

            if (answer === null) return null;

            const said = answer.trim();

            if (!said) {
                if (televisions.length === 1) return televisions[0].ip;
                if (!televisions.length) return null;

                ui.warn(`Pick one, 1 to ${televisions.length}, or type an address.`);
                continue;
            }

            if (televisions.length && /^\d{1,3}$/.test(said)) {
                const chosen = Number(said);

                if (chosen >= 1 && chosen <= televisions.length) return televisions[chosen - 1].ip;

                ui.warn(`There is no ${chosen} in that list — pick 1 to ${televisions.length}, or type an address.`);
                continue;
            }

            if (!isAddress(said)) {
                ui.warn(`${said} is not an address.`);
                continue;
            }

            ui.note(ui.style.dim(`  asking ${said}...`));

            const device = await describe(said, 5000);

            if (!device) {
                ui.warn(`Nothing answered on ${said}:8001 — is that the TV, and is it on?`);
                continue;
            }

            ui.ok('found', `${nameOf(device)} \u00b7 ${detailOf(device)}`);

            return said;
        }
    } finally {
        rl.close();
    }
};

// The address, or null when there is nobody to ask — the caller says what to do about that.
const choose = async () => {
    if (!process.stdin.isTTY) return null;

    const where = subnets().prefixes.map((prefix) => `${prefix}.0/24`).join(', ') || 'this machine';

    ui.note(`  ${ui.style.dim(`looking for televisions on ${where}...`)}`);

    const { televisions } = await sweep();

    if (!televisions.length) {
        ui.fail('no televisions', ui.style.dim(where));
        ui.blank();
        ui.note('  Nothing answered. A set has to be on, and on the same network as this');
        ui.note('  machine — a guest network or a VPN is enough to hide it.');
    } else {
        ui.ok(televisions.length === 1 ? 'found a television' : `found ${televisions.length} televisions`, where);
        ui.blank();
        listed(televisions);
    }

    return ask(televisions);
};

module.exports = { choose, sweep, isAddress, nameOf, modeOf, detailOf };

File: installer/internal/discover/discover.go
// Package discover finds Samsung televisions on the local network.
//
// Every Samsung set answers http://<ip>:8001/api/v2/ with its model and developer-mode state, so a
// sweep of the /24s this machine sits in finds them without sending anyone to their router. Asking
// with HTTP costs far more than the network does — a /24 takes about ten seconds that way and one
// and a half opened by hand — so the sweep knocks on the port first and only asks the few addresses
// that answered what they are.
package discover

import (
	"context"
	"encoding/json"
	"encoding/xml"
	"fmt"
	"net"
	"net/http"
	"sort"
	"strings"
	"sync"
	"time"
)

// DevicePort is the Samsung device API, open on every set whether or not developer mode is on.
const DevicePort = 8001

const (
	knockTimeout = 1500 * time.Millisecond
	askTimeout   = 5 * time.Second
	atOnce       = 254
)

// TV is one television that answered.
type TV struct {
	IP          string
	Name        string
	Model       string
	DeveloperOn bool
	DeveloperIP string
}

type deviceAPI struct {
	Device struct {
		Name          string `json:"name"`
		ModelName     string `json:"modelName"`
		Model         string `json:"model"`
		DeveloperMode string `json:"developerMode"`
		DeveloperIP   string `json:"developerIP"`
	} `json:"device"`
}

// Subnets returns the /24 prefixes this machine sits in, and its own addresses within them.
func Subnets() (prefixes []string, mine []string) {
	interfaces, err := net.Interfaces()
	if err != nil {
		return nil, nil
	}

	seen := map[string]bool{}

	for _, item := range interfaces {
		if item.Flags&net.FlagUp == 0 || item.Flags&net.FlagLoopback != 0 {
			continue
		}

		addresses, err := item.Addrs()
		if err != nil {
			continue
		}

		for _, address := range addresses {
			ipNet, ok := address.(*net.IPNet)
			if !ok || ipNet.IP.To4() == nil {
				continue
			}

			ip := ipNet.IP.To4().String()
			mine = append(mine, ip)

			prefix := ip[:strings.LastIndex(ip, ".")]
			if !seen[prefix] {
				seen[prefix] = true
				prefixes = append(prefixes, prefix)
			}
		}
	}

	return prefixes, mine
}

// knock opens a socket and closes it: the cheapest question that separates 254 dead addresses from
// the handful worth asking properly.
func knock(ctx context.Context, ip string) bool {
	dialer := net.Dialer{Timeout: knockTimeout}

	conn, err := dialer.DialContext(ctx, "tcp", fmt.Sprintf("%s:%d", ip, DevicePort))
	if err != nil {
		return false
	}

	_ = conn.Close()

	return true
}

// Describe asks one address what it is. Something else may sit on 8001, so an answer that does not
// look like a television is not one rather than an error to report.
func Describe(ctx context.Context, ip string) (*TV, error) {
	ctx, cancel := context.WithTimeout(ctx, askTimeout)
	defer cancel()

	url := fmt.Sprintf("http://%s:%d/api/v2/", ip, DevicePort)

	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}

	response, err := http.DefaultClient.Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	var answer deviceAPI
	if err := json.NewDecoder(response.Body).Decode(&answer); err != nil {
		return nil, err
	}

	device := answer.Device
	if device.ModelName == "" && device.Model == "" && device.Name == "" {
		return nil, fmt.Errorf("%s answered on %d but is not a television", ip, DevicePort)
	}

	return &TV{
		IP:          ip,
		Name:        plain(pick(device.Name, device.ModelName, device.Model)),
		Model:       device.ModelName,
		DeveloperOn: device.DeveloperMode == "1",
		DeveloperIP: device.DeveloperIP,
	}, nil
}

// Sweep knocks on every address in the local /24s and describes the ones that answered.
func Sweep(ctx context.Context) (televisions []TV, prefixes []string) {
	prefixes, mine := Subnets()

	ours := map[string]bool{}
	for _, ip := range mine {
		ours[ip] = true
	}

	var addresses []string
	for _, prefix := range prefixes {
		for host := 1; host < 255; host++ {
			if ip := fmt.Sprintf("%s.%d", prefix, host); !ours[ip] {
				addresses = append(addresses, ip)
			}
		}
	}

	var (
		wait      sync.WaitGroup
		guard     = make(chan struct{}, atOnce)
		mutex     sync.Mutex
		listening []string
	)

	for _, ip := range addresses {
		wait.Add(1)

		go func(ip string) {
			defer wait.Done()

			guard <- struct{}{}
			defer func() { <-guard }()

			if knock(ctx, ip) {
				mutex.Lock()
				listening = append(listening, ip)
				mutex.Unlock()
			}
		}(ip)
	}

	wait.Wait()

	for _, ip := range listening {
		wait.Add(1)

		go func(ip string) {
			defer wait.Done()

			tv, err := Describe(ctx, ip)
			if err != nil {
				return
			}

			mutex.Lock()
			televisions = append(televisions, *tv)
			mutex.Unlock()
		}(ip)
	}

	wait.Wait()

	sort.Slice(televisions, func(a, b int) bool {
		return numeric(televisions[a].IP) < numeric(televisions[b].IP)
	})

	return televisions, prefixes
}

func pick(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}

	return "Samsung TV"
}

// A set named `65" OLED` reports it as `65&quot; OLED`.
func plain(text string) string {
	var holder struct {
		Value string `xml:",chardata"`
	}

	if err := xml.Unmarshal([]byte("<t>"+text+"</t>"), &holder); err == nil {
		return holder.Value
	}

	return text
}

func numeric(ip string) uint32 {
	parsed := net.ParseIP(ip).To4()
	if parsed == nil {
		return 0
	}

	return uint32(parsed[0])<<24 | uint32(parsed[1])<<16 | uint32(parsed[2])<<8 | uint32(parsed[3])
}

// PrimaryAddress is the address this machine would use to reach the outside world, which on a
// machine with a virtual-machine bridge or a VPN is the only one of several worth typing into a
// television. Connecting a UDP socket sends nothing; it just asks the routing table to choose.
func PrimaryAddress() string {
	conn, err := net.Dial("udp", "8.8.8.8:80")
	if err != nil {
		return ""
	}
	defer conn.Close()

	address, ok := conn.LocalAddr().(*net.UDPAddr)
	if !ok {
		return ""
	}

	return address.IP.String()
}
